src/solvers/refinement/string_constraint_generator_concat.cpp

```diff
@@ -11,44 +11,60 @@ Author: Romain Brenguier, romain.brenguier@diffblue.com
 
 /*******************************************************************\
 
-Function: string_constraint_generatort::add_axioms_for_concat
+Function: string_constraint_generatort::add_axioms_for_concat_substr
 
-  Inputs: two string expressions
+  Inputs:
+    s1 - string expression
+    s2 - string expression
+    start_index - expression representing an integer
+    end_index - expression representing an integer
 
  Outputs: a new string expression
 
  Purpose: add axioms to say that the returned string expression is equal to
-          the concatenation of the two string expressions given as input
+          the concatenation of s1 with the substring of s2 starting at index
+          start_index and ending at index end_index.
+          If start_index >= end_index the value returned is s1.
+
+          TODO: For now it just assumes end_index < s2.length. This
+          should be handled in the Java model.
 
 \*******************************************************************/
 
-string_exprt string_constraint_generatort::add_axioms_for_concat(
-  const string_exprt &s1, const string_exprt &s2)
+string_exprt string_constraint_generatort::add_axioms_for_concat_substr(
+  const string_exprt &s1,
+  const string_exprt &s2,
+  const exprt &start_index,
+  const exprt &end_index)
 {
   const refined_string_typet &ref_type=to_refined_string_type(s1.type());
   string_exprt res=fresh_string(ref_type);
 
   // We add axioms:
-  // a1 : |res|=|s1|+|s2|
-  // a2 : |s1| <= |res| (to avoid overflow with very long strings)
-  // a3 : |s2| <= |res| (to avoid overflow with very long strings)
-  // a4 : forall i<|s1|. res[i]=s1[i]
-  // a5 : forall i<|s2|. res[i+|s1|]=s2[i]
-
-  equal_exprt a1(
-    res.length(), plus_exprt_with_overflow_check(s1.length(), s2.length()));
+  // a1 : end_index > start_index => |res|=|s1|+ end_index - start_index
+  // a2 : end_index <= start_index => res = s1
+  // a3 : forall i<|s1|. res[i]=s1[i]
+  // a4 : forall i< end_index - start_index. res[i+|s1|]=s2[start_index+i]
```

Comment:
It currently just appends the empty string, so we do get a generated test. The exception throw will then be covered in the models.